CoreScriptsRoot/Modules/Panel3D.lua

```diff
@@ -120,42 +244,69 @@ end
+local savedMouseBehavior = Enum.MouseBehavior.Default
 function Panel3D.OnRenderStep()
+	if not UserInputService.VREnabled then
+		return
+	end
 	local cameraRenderCFrame = workspace.CurrentCamera:GetRenderCFrame()
+	local userHeadCFrame = UserInputService:GetUserCFrame(Enum.UserCFrame.Head)
 	local cameraLook = cameraRenderCFrame.lookVector
 	local cameraHorizontalVector = Vector3.new(cameraLook.X, 0, cameraLook.Z).unit
 	local cameraPitchAngle = math.asin(cameraLook.Y)
 
 	local position = workspace.CurrentCamera.CFrame.p
-	local panelsOrigin = CFrame.new(position) * baseHorizontal
+	local panelsOrigin = CFrame.new(position) * baseHorizontal * CFrame.new(basePosition)
+
+	local isAboveThreshold = false
+	if cameraPitchAngle > panelLockThreshold or resetOrientationFlag then
+		isAboveThreshold = true
+		baseHorizontal = CFrame.new(zeroVector, cameraHorizontalVector)
+		basePosition = userHeadCFrame.p
+		resetOrientationFlag = false
+	end
+
+	local anyModal = false
 	for panelId, panel in pairs(panels) do
-		local showPanel = true
-		if cameraPitchAngle > panelLockThreshold[panelId] then
-			baseHorizontal = CFrame.new(zeroVector, cameraHorizontalVector)
-			showPanel = false
+		if panel.visibilityBehavior == Panel3D.Visibility.BelowAngleThreshold then
+			panel.visible = not isAboveThreshold and not modalPanelIsOpen
 		end
-		if not cursor.Visible or menuOpened then
-			showPanel = false
+		if panel.visibilityBehavior == Panel3D.Visibility.Modal then
+			if panel.visible then
+				anyModal = true
+			end
+		else
+			if anyModal or modalPanelIsOpen then
+				panel.visible = false
+			end
 		end
 
-		local panelPosition = panelsOrigin:pointToWorldSpace(panel.vector)
-		local panelCFrame = CFrame.new(panelPosition, panelsOrigin.p)
-		panel.part.CFrame = panelCFrame
-
-		local toPanel = (panelCFrame.p - cameraRenderCFrame.p).unit
-
-		local transparency = showPanel and 1 - (math.max(0, cameraLook:Dot(toPanel)) ^ panelTransparencyBias[panelId]) or 1
-		for _, callback in pairs(panel.transparencyCallbacks) do
-			callback(transparency)
-		end
-		if not showPanel then
+		if not panel.visible then
 			panel.part.Parent = nil
 		else
 			panel.part.Parent = workspace.CurrentCamera --TODO: move to new 3D gui space
+
+			local panelCFrame;
+			if panel.orientationMode == Panel3D.Orientation.Fixed and panel.orientation then
+				panel.part.CFrame = workspace.CurrentCamera.CFrame * panel.orientation
+				panelCFrame = panel.part.CFrame
+			else
+				local panelPosition = panelsOrigin:pointToWorldSpace(panel.vector)
+				panelCFrame = CFrame.new(panelPosition, panelsOrigin.p)
+				panel.part.CFrame = panelCFrame
+			end
+
+			local toPanel = (panelCFrame.p - cameraRenderCFrame.p).unit
+
+			local transparency = panel.visible and 1 - (math.max(0, cameraLook:Dot(toPanel)) ^ panelTransparencyBias[panelId]) or 1
+			for _, callback in pairs(panel.transparencyCallbacks) do
+				callback(transparency)
+			end
 		end
 	end
+	modalPanelIsOpen = anyModal
```

Comment:
Instead of doing this intensive for loop every frame, we should try to update `modalPanelIsOpen` in `panel:SetVisible(visible)` and `panel:SetModal()`